### ISSUE TITLE
properties are not being copied over correctly

### ISSUE THREAD

**Issue by yoshuawuyts** [yoshuawuyts/nanomorph#6]
e.g. we're evaporating them hah

**Comment by kristoferjoseph** [comment]
Hey bud, 
Do you have an example of this so I could fix it?
I kind of live to make failing tests pass 😜

**Comment by yoshuawuyts** [comment]
Ooph, this has been a while, but I believe a few rounds of mutation on any arbitrary tree would fail; e.g.
```js
var el1 = html`
  <div class="bar">hey</div>
`

var el2 = html`
  <div>mate</div>
`

var el = morph(el1, el2)
morph(el, el1) // at this point values would stop to be copied over correctly
```

Though thinking about it, it could be a bug because `el1` is being mutated and now has a different value. I think after #12 and #20 are in we should test it in some random apps and see if this bug pops up (:

**Comment by kristoferjoseph** [comment]
Totally.
That was what I was thinking... swap out for `nanomorph` in a demo app I have and see if anything gets weird.

**Comment by yoshuawuyts** [comment]
okay, there's definitely weird things going on. Did a quick port of `nanomorph` to choo and it looks like things aren't like they should be hmm - [here's the wip branch](https://github.com/yoshuawuyts/choo/compare/nanomorph-pt2?expand=1)

<img width="1278" alt="screen shot 2017-01-07 at 01 55 28" src="https://cloud.githubusercontent.com/assets/2467194/21737751/76506a7c-d47c-11e6-904c-1ff1f37c51b5.png">

**Comment by kristoferjoseph** [comment]
This demo seems to be working fine: https://gist.github.com/kristoferjoseph/bea5f4957a84f1583534a22829c6b4ea

What am I doing wrong?

**Comment by kristoferjoseph** [comment]
@yoshuawuyts looks like you have a typo in your require 
https://github.com/yoshuawuyts/choo/blob/764ebe1c7d110fad7c2de8548b3b12a471c974c2/mount.js#L3

s/update-dome/update-dome/

**Comment by yoshuawuyts** [comment]
@kristoferjoseph neato; great catch! Unforch we weren't using it in the example so it's not the cause of the issue - did you try running the example tho? Something's definitely up.

**Comment by kristoferjoseph** [comment]
Bummer 😞 
I did try to run the example, but it threw a bunch of errors I didn't want to fix so I gave up 😬

**Comment by rreusser** [comment]
Does this include events? If so, there's good news and bad news. [This issue](https://github.com/maxogden/yo-yo/issues/56) was a bit of a blocker with yo-yo because it seemed the UI wasn't updating to match the DOM, which is just frustrating and fishy. I swapped nanomorph in for yo-yo and [this is the result](http://codepen.io/rsreusser/pen/QdjyQX?editors=1010). The bad news is that now it's the exact opposite: the input does not update after you've dragged it. The good news is that this seems less browser-quirky and more like an event handler on the number input that's maybe just not getting copied over. I can file a separate issue, but seems like maybe it falls under the same umbrella.

**Comment by yoshuawuyts** [comment]
@kj sorry to hear; yeah we gotta clean those examples up, but probs not
gonna do that today (: - thinking of removing them all and only maintaing a
todomvc app in a separate repo - also useful for benchmarks and stuff

**Comment by kristoferjoseph** [comment]
@rreusser yup.
Was just discussing this earlier on a different issue.
The code only overwrites the value of an input if it doesn't already have one.
This is to keep updates from losing user input.
It is making sure to maintain the users value in case an unrelated update ( in this case changing the value of a different control ) does not clobber the users action.

**Comment by yoshuawuyts** [comment]
Looks like we got issues with the sibling nodes / updates on the same node going still. This is using the latest version of `nanomorph`.

![kapture 2017-01-12 at 19 58 27](https://cloud.githubusercontent.com/assets/2467194/21903843/9ca2bfc0-d901-11e6-8efb-211643bd953c.gif)

[branch on choo](https://github.com/yoshuawuyts/choo/compare/nanomorph-pt2?expand=1)

**Comment by rreusser** [comment]
Is that preexisting or a regression? I'm not 100% sure the actual algo react et al use to match by `key`, but I'm quiet certain my addition, though an incremental improvement, isn't there yet. Seems necessary to have some sort of lookup table and actually manage rearranging nodes by `id` or something. Is that where the weakmap comes in?

**Comment by kristoferjoseph** [comment]
@yoshuawuyts Ya, hmmm.
Makes me want to take a second look at `firstElement` etc. to see if we can use a recursive walk function instead of a for loop. 
[something like this???](https://gist.github.com/kristoferjoseph/c866f9946fd5e3c881c293b8948c061b)
@rreusser you are right, we should look into keying the tree somehow.
